receiver/flinkmetricsreceiver/process.go

```diff
@@ -0,0 +1,220 @@
+				_ = s.mb.RecordFlinkJobCheckpointCountDataPoint(now, metric.Value, metadata.AttributeCheckpointCompleted)
+			case "numberOfFailedCheckpoints":
+				_ = s.mb.RecordFlinkJobCheckpointCountDataPoint(now, metric.Value, metadata.AttributeCheckpointFailed)
+			}
+		}
+	}
+	if len(jobsMetricsInstances) > 0 {
+		s.mb.EmitForResource(
+			metadata.WithHostName(jobsMetricsInstances[0].Host),
+			metadata.WithFlinkJobName(jobsMetricsInstances[0].JobName),
+		)
+	}
+}
+
+func (s *flinkmetricsScraper) processSubtaskMetrics(now pcommon.Timestamp, subtaskMetricsInstances []*models.SubtaskMetrics) {
+	for _, subtaskMetrics := range subtaskMetricsInstances {
+		for _, metric := range subtaskMetrics.Metrics {
+			switch {
+			// record task metrics
+			case metric.ID == "numRecordsIn":
+				_ = s.mb.RecordFlinkTaskRecordCountDataPoint(now, metric.Value, metadata.AttributeRecordIn)
+			case metric.ID == "numRecordsOut":
+				_ = s.mb.RecordFlinkTaskRecordCountDataPoint(now, metric.Value, metadata.AttributeRecordOut)
+			case metric.ID == "numLateRecordsDropped":
+				_ = s.mb.RecordFlinkTaskRecordCountDataPoint(now, metric.Value, metadata.AttributeRecordDropped)
+			}
+		}
+		s.mb.EmitForResource(
+			metadata.WithHostName(subtaskMetrics.Host),
+			metadata.WithFlinkTaskmanagerID(subtaskMetrics.TaskmanagerID),
+			metadata.WithFlinkJobName(subtaskMetrics.JobName),
+			metadata.WithFlinkTaskName(subtaskMetrics.TaskName),
+			metadata.WithFlinkSubtaskIndex(subtaskMetrics.SubtaskIndex),
+		)
+	}
+
+	for _, subtaskMetrics := range subtaskMetricsInstances {
+		for _, metric := range subtaskMetrics.Metrics {
+			switch {
+			// record operator metrics
+			case strings.Contains(metric.ID, ".numRecordsIn"):
+				operatorName := strings.Split(metric.ID, ".numRecordsIn")
+				_ = s.mb.RecordFlinkOperatorRecordCountDataPoint(now, metric.Value, operatorName[0], metadata.AttributeRecordIn)
+			case strings.Contains(metric.ID, ".numRecordsOut"):
+				operatorName := strings.Split(metric.ID, ".numRecordsOut")
+				_ = s.mb.RecordFlinkOperatorRecordCountDataPoint(now, metric.Value, operatorName[0], metadata.AttributeRecordOut)
+			case strings.Contains(metric.ID, ".numLateRecordsDropped"):
+				operatorName := strings.Split(metric.ID, ".numLateRecordsDropped")
+				_ = s.mb.RecordFlinkOperatorRecordCountDataPoint(now, metric.Value, operatorName[0], metadata.AttributeRecordDropped)
+			case strings.Contains(metric.ID, ".currentOutputWatermark"):
+				operatorName := strings.Split(metric.ID, ".currentOutputWatermark")
+				_ = s.mb.RecordFlinkOperatorWatermarkOutputDataPoint(now, metric.Value, operatorName[0])
+			}
+		}
+		s.mb.EmitForResource(
+			metadata.WithHostName(subtaskMetrics.Host),
+			metadata.WithFlinkTaskmanagerID(subtaskMetrics.TaskmanagerID),
+			metadata.WithFlinkJobName(subtaskMetrics.JobName),
+			metadata.WithFlinkTaskName(subtaskMetrics.TaskName),
+			metadata.WithFlinkSubtaskIndex(subtaskMetrics.SubtaskIndex),
+		)
+	}
```

Comment:
Why is this split into two loops? You're emitting twice for each resource?

Reply:
it can be combined now. updated